### PR TITLE
Configuration Manager

### PR DESCRIPTION
**Pull request by gleblebedev**

### CODE REVIEW

File: Source/Urho3D/Engine/ConfigManager.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
+//
+// Copyright (c) 2022-2022 the rbfx project.
```
Comment (rokups) [inline Source/Urho3D/Engine/ConfigManager.h:2]: Starting copyright year is 2017.

File: Source/Urho3D/Engine/ConfigManager.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
+//
+// Copyright (c) 2017-2020 the rbfx project.
```
Comment (rokups) [inline Source/Urho3D/Engine/ConfigManager.cpp:2]: and ending copyright year is 2022 😄

File: Source/Urho3D/Core/Context.cpp [hunk 3]
```diff
@@ -159,10 +159,11 @@ Context::Context()
 
 Context::~Context()
 {
+#ifndef MINI_URHO
+    // Save any configuration that may be unsaved at runtime.
     auto* configManager = GetSubsystem<ConfigManager>();
     configManager->SaveAll();
```
Comment (rokups) [inline Source/Urho3D/Core/Context.cpp:165]: Why is this excluded from minimal builds? I can see config manager being useful for tools.

Comment (gleblebedev) [reply Source/Urho3D/Core/Context.cpp:165]: It fails to compile. I thought that it is by design.

Comment (rokups) [reply Source/Urho3D/Core/Context.cpp:165]: Why would it fail though? `MINI_URHO` is there to provide core utilities to tools, so it avoids compiling graphics/physics/few other things, but does include xml/json. There should be nothing that prevents it to build i think. SDL may also be omitted, not sure. But if it is - this class would still be useful when config path is explicitly specified.

File: Source/Urho3D/Engine/ConfigManager.cpp [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    const char* configSubfoler = "Config";
+    const auto fileSystem = context_->GetSubsystem<FileSystem>();
+#if defined(__ANDROID__) || defined(__IOS__)
+    configurationFolder_ = fileSystem->GetUserDocumentsDir() + configSubfoler;
+#elif defined(__EMSCRIPTEN__)
+    // TODO: implement local storage utilization for web
+    configurationFolder_ = configSubfoler;
+#else
+    configurationFolder_ = fileSystem->GetProgramDir() + configSubfoler;
+#endif
```
[the remark this answers is not in the document]
Comment (gleblebedev) [reply Source/Urho3D/Engine/ConfigManager.cpp]: It doesn't work on UWP.

"Can't save file C:/Users/gleb/AppData/Local/Packages/39ad144a-075c-4e1b-9d81-1a1e4672a258_zzcskrky2cb1y/LocalState/Urho3D Rebel Fork/Unspecified Application/configs/SampleConfigFile.json"

[the remark this answers is not in the document]
Comment (gleblebedev) [reply Source/Urho3D/Engine/ConfigManager.cpp]: Android works

**Comment by eugeneko** [comment]
I get a feeling that this ConfigManager is overengineered and overly complicated. Let's discuss.

**Comment by eugeneko** [comment]
Okay, so, what I think about this stuff

- We need new class like `UserFile`, which is very similar to `File`, is inherited from `AbstractFile`, and uses `File` under the hood.
  - Name of this file is relative to the user directory (i.e. app pref)
  - On Web, this `UserFile` will use `localStorage` instead.
  - Later, we can integrate it into `ResourceCache`
  - This way we will support *any* user files, not just "configurations"
- In order to configure stuff like fullscreen flag, we need to load some predefined user file in Player.exe
  - This is the place: https://github.com/rbfx/rbfx/blob/master/Source/Player/Player.cpp#L57-L59
  - We just load `/LaunchParameters.json` from user storage if it exists, and apply loaded parameters to `engineParameters_`

**Comment by rokups** [comment]
> We need new class like UserFile

Do we really? What is wrong with just creating a `File` in `fs->GetAppPreferencesDir(..)`? Admittedly `GetAppPreferencesDir()` could be improved to fetch default values when parameters are not specified. One more distinct file type that does nothing other than restricting where file is stored sounds like different kind of overengineering

**Comment by eugeneko** [comment]
> Do we really? What is wrong with just creating a File in fs->GetAppPreferencesDir(..)?

Because there are no writeable Files in Web, and we will have to implement UserFile in Web from scratch.

**Comment by rokups** [comment]
Im pretty sure web can have writable files. browsers have a local storage that can be used

**Comment by eugeneko** [comment]
And to use local storage we will need to reimplement `File` from scartch. We either put it into `File` itself, making it 5 times more complicated, or put it outside.

**Comment by rokups** [comment]
I was hoping SDL would abstract it for us 👀

**Pull request closed without merging**